cpu/stm32/periph/eth.c

```diff
@@ -435,8 +512,19 @@ void stm32_eth_isr_eth_wkup(void)
     cortexm_isr_end();
 }
 
+int stm32_eth_get_rx_status_owned(void)
+{
+    return (!(rx_curr->status & DESC_OWN));
+}
+
 static void stm32_eth_isr(netdev_t *netdev) {
-    netdev->event_callback(netdev, NETDEV_EVENT_RX_COMPLETE);
+    if (stm32_eth_get_rx_status_owned()) {
+        netdev->event_callback(netdev, NETDEV_EVENT_RX_COMPLETE);
+    }
+    if (_link_status && _first_call) {
+        netdev->event_callback(netdev, NETDEV_EVENT_LINK_UP);
+        _first_call = 0;
+    }
```

Comment:
No need to call `stm32_eth_get_rx_status_owned()` here. If there is no notification for link state changes pending, the ISR can only be caused by an RX event. And if the pseudo module for link status polling is not used, it will always be an RX event.

You could go as simple as:

```C
#ifdef MODULE_STM32_ETH_LINK_UP
    switch (link_state) {
    case LINK_STATE_UP:
        netdev->event_callback(netdev, NETDEV_EVENT_LINK_UP);
        link_state = LINK_STATE_NOTIFIED_UP;
        return;
    case LINK_STATE_DOWN:
        netdev->event_callback(netdev, NETDEV_EVENT_LINK_DOWN);
        link_state = LINK_STATE_NOTIFIED_DOWN;
        return;
    }
#endif /* MODULE_STM32_ETH_LINK_UP */

    netdev->event_callback(netdev, NETDEV_EVENT_RX_COMPLETE);
```